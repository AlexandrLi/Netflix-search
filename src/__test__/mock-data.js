export const movies = [{
    vote_count: 8410,
    id: 122,
    video: false,
    vote_average: 8.2,
    title: 'The Lord of the Rings: The Return of the King',
    popularity: 22.480443,
    poster_path: '/uexxR7Kw1qYbZk0RYaF9Rx5ykbj.jpg',
    original_language: 'en',
    original_title: 'The Lord of the Rings: The Return of the King',
    genre_ids: [
      12,
      14,
      28
    ],
    backdrop_path: '/8BPZO0Bf8TeAy8znF43z8soK3ys.jpg',
    adult: false,
    overview: 'Aragorn is revealed as the heir to the ancient kings as he, Gandalf and the other members of the broken fellowship struggle to save Gondor from Sauron\'s forces. Meanwhile, Frodo and Sam bring the ring closer to the heart of Mordor, the dark lord\'s realm.',
    release_date: '2003-12-01'
  },
  {
    vote_count: 9090,
    id: 120,
    video: false,
    vote_average: 8.1,
    title: 'The Lord of the Rings: The Fellowship of the Ring',
    popularity: 36.635241,
    poster_path: '/bxVxZb5O9OxCO0oRUNdCnpy9NST.jpg',
    original_language: 'en',
    original_title: 'The Lord of the Rings: The Fellowship of the Ring',
    genre_ids: [
      12,
      14,
      28
    ],
    backdrop_path: '/pIUvQ9Ed35wlWhY2oU6OmwEsmzG.jpg',
    adult: false,
    overview: 'Young hobbit Frodo Baggins, after inheriting a mysterious ring from his uncle Bilbo, must leave his home in order to keep it from falling into the hands of its evil creator. Along the way, a fellowship is formed to protect the ringbearer and make sure that the ring arrives at its final destination: Mt. Doom, the only place where it can be destroyed.',
    release_date: '2001-12-18'
  },
  {
    vote_count: 7842,
    id: 121,
    video: false,
    vote_average: 8,
    title: 'The Lord of the Rings: The Two Towers',
    popularity: 24.591994,
    poster_path: '/wf3v0Pn09jnT5HSaYal7Ami3bdA.jpg',
    original_language: 'en',
    original_title: 'The Lord of the Rings: The Two Towers',
    genre_ids: [
      12,
      14,
      28
    ],
    backdrop_path: '/dG4BmM32XJmKiwopLDQmvXEhuHB.jpg',
    adult: false,
    overview: 'Frodo and Sam are trekking to Mordor to destroy the One Ring of Power while Gimli, Legolas and Aragorn search for the orc-captured Merry and Pippin. All along, nefarious wizard Saruman awaits the Fellowship members at the Orthanc Tower in Isengard.',
    release_date: '2002-12-18'
  }
]

export const yodaMovies = [{
    vote_count: 1,
    id: 371796,
    video: false,
    vote_average: 7,
    title: 'The New Yoda Chronicles: Escape from the Jedi Temple',
    popularity: 1.677475,
    poster_path: '/2nwar7SCeY9fktHRA8LFwHRBqiI.jpg',
    original_language: 'en',
    original_title: 'The New Yoda Chronicles: Escape from the Jedi Temple',
    genre_ids: [
      16,
      10751
    ],
    backdrop_path: null,
    adult: false,
    overview: '',
    release_date: '2014-12-09'
  },
  {
    vote_count: 1,
    id: 424492,
    video: false,
    vote_average: 7,
    title: 'The New Yoda Chronicles: Clash of the Skywalkers',
    popularity: 1.053418,
    poster_path: '/d0nFx7eQvXsGZgRkVLY7NymtZXd.jpg',
    original_language: 'en',
    original_title: 'The New Yoda Chronicles: Clash of the Skywalkers',
    genre_ids: [
      16
    ],
    backdrop_path: null,
    adult: false,
    overview: 'The enraged Emperor sends Darth Vader and his entire fleet to find and destroy Luke as revenge for his destruction of the Holocrons. Luke, however, finds inspiration from the last remaining Holocron -- a daring move pulled off years earlier by his father, Anakin. The emboldened Luke uses Anakin\'s daring move to lure Vader and the Imperial Fleet into a trap that could turn the tide for the Rebellion - or lead to disaster.',
    release_date: '2014-11-23'
  }
]

export const movie = {
  adult: false,
  backdrop_path: '/fUjPnCmmBVHLNdgz2sqb1zp39Pr.jpg',
  belongs_to_collection: {
    id: 1241,
    name: 'Harry Potter Collection',
    poster_path: '/eVPs2Y0LyvTLZn6AP5Z6O2rtiGB.jpg',
    backdrop_path: '/wfnMt6LGqYHcNyOfsuusw5lX3bL.jpg'
  },
  budget: 125000000,
  genres: [{
      id: 12,
      name: 'Adventure'
    },
    {
      id: 14,
      name: 'Fantasy'
    },
    {
      id: 10751,
      name: 'Family'
    }
  ],
  homepage: 'http://harrypotter.warnerbros.com/harrypotterandthedeathlyhallows/mainsite/index.html',
  id: 671,
  imdb_id: 'tt0241527',
  original_language: 'en',
  original_title: 'Harry Potter and the Philosopher\'s Stone',
  overview: 'Harry Potter has lived under the stairs at his aunt and uncle\'s house his whole life. But on his 11th birthday, he learns he\'s a powerful wizard -- with a place waiting for him at the Hogwarts School of Witchcraft and Wizardry. As he learns to harness his newfound powers with the help of the school\'s kindly headmaster, Harry uncovers the truth about his parents\' deaths -- and about the villain who\'s to blame.',
  popularity: 32.913022,
  poster_path: '/dCtFvscYcXQKTNvyyaQr2g2UacJ.jpg',
  production_companies: [{
      name: '1492 Pictures',
      id: 436
    },
    {
      name: 'Warner Bros.',
      id: 6194
    },
    {
      name: 'Heyday Films',
      id: 7364
    }
  ],
  production_countries: [{
      iso_3166_1: 'GB',
      name: 'United Kingdom'
    },
    {
      iso_3166_1: 'US',
      name: 'United States of America'
    }
  ],
  release_date: '2001-11-16',
  revenue: 976475550,
  runtime: 152,
  spoken_languages: [{
    iso_639_1: 'en',
    name: 'English'
  }],
  status: 'Released',
  tagline: 'Let the Magic Begin.',
  title: 'Harry Potter and the Philosopher\'s Stone',
  video: false,
  vote_average: 7.5,
  vote_count: 7441
};

export const anotherMovie = {
  vote_count: 9090,
  id: 120,
  video: false,
  vote_average: 8.1,
  title: 'The Lord of the Rings: The Fellowship of the Ring',
  popularity: 36.635241,
  poster_path: '/bxVxZb5O9OxCO0oRUNdCnpy9NST.jpg',
  original_language: 'en',
  original_title: 'The Lord of the Rings: The Fellowship of the Ring',
  genre_ids: [
    12,
    14,
    28
  ],
  backdrop_path: '/pIUvQ9Ed35wlWhY2oU6OmwEsmzG.jpg',
  adult: false,
  overview: 'Young hobbit Frodo Baggins, after inheriting a mysterious ring from his uncle Bilbo, must leave his home in order to keep it from falling into the hands of its evil creator. Along the way, a fellowship is formed to protect the ringbearer and make sure that the ring arrives at its final destination: Mt. Doom, the only place where it can be destroyed.',
  release_date: '2001-12-18'
}

export const FAKE_ACTION = 'FAKE_ACTION';

export const initialState = {
  movies: [],
  selectedMovie: null,
  sortBy: 'release'
};