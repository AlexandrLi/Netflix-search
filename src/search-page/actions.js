const API_KEY_PARAM = 'api_key=e2c6b28a24bafd10044b1fc5397cc4d2';
const ROOT_URL = `https://api.themoviedb.org/3/search`;

export const FETCH_MOVIES = 'FETCH_MOVIES';
export const CHANGE_SEARCH_TYPE = 'CHANGE_SEARCH_TYPE';
export const CHANGE_SORT_TYPE = 'CHANGE_SORT_TYPE';

function fetchMoviesSuccess(movies) {
  return {
    type: FETCH_MOVIES,
    payload: movies
  }
}

export function fetchMovies(query) {
  return (dispatch, getState) => {
    const { searchType, sortBy } = getState();
    const url = `${ROOT_URL}/${searchType}?${API_KEY_PARAM}&query=${query}`;
    fetch(url)
      .then(response => response.json())
      .then(response => {
        if (searchType === 'movie') {
          dispatch(fetchMoviesSuccess(sort(sortBy, response.results)));
        } else {
          dispatch(fetchMoviesSuccess(sort(sortBy, response.results.map(movie => {
            return {
              id: movie.id,
              title: movie.name,
              release_date: movie.first_air_date,
              poster_path: movie.poster_path,
              vote_average: movie.vote_average
            }
          }))));
        }
      });
  }
}

export function changeSearchType(type) {
  return {
    type: CHANGE_SEARCH_TYPE,
    payload: type
  }
}

export function changeSortType(type) {
  return {
    type: CHANGE_SORT_TYPE,
    payload: type
  }
}

function sort(type, data) {
  if (type === 'release') {
    return data.sort((item1, item2) => {
      if (!item1.release_date) {
        return 1;
      } else {
        return (Date.parse(item2.release_date) - Date.parse(item1.release_date))
      }
    })
  } else {
    return data.sort((item1, item2) => { return (item2.vote_average - item1.vote_average) })
  }
}

export function sortMovies(type) {
  return (dispatch, getState) => {
    const { movies } = getState();
    dispatch(changeSortType(type));
    dispatch(fetchMoviesSuccess(sort(type, movies)));
  }
}