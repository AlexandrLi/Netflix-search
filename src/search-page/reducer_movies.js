export default function (state = [], action) {
  console.log(action.payload);
  switch (action.type) {
    case 'FETCH_MOVIES':
      return [...state, action.payload];
    default:
      return state;
  }
}