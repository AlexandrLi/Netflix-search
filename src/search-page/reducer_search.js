export default function (state = 'movie', action) {
  switch (action.type) {
    case 'CHANGE_SEARCH_TYPE':
      return action.payload;
    default:
      return state;
  }
}