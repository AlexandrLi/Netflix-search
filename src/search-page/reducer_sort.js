export default function (state = 'RELEASE_DATE', action) {
  console.log(action.payload);
  switch (action.type) {
    case 'CHANGE_SORT_TYPE':
      return action.payload;
    default:
      return state;
  }
}