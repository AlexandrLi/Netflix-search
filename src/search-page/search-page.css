header {
  display: flex;
  flex-direction: column;
  padding: 0 100px;
  height: 280px;
  color: white;
  position: relative;
  background: rgba(0, 0, 0, 0.75);
  overflow: hidden;
}

header:before {
  content: "";
  height: 100%;
  width: 100%;
  position: absolute;
  z-index: -1;
  background: url('https://assets.nflxext.com/ffe/siteui/vlv3/dc18d1c3-9e06-41a4-a5b0-409388e3f9e8/532f465f-9176-4555-8e82-11ec6e34201e/KZ-en-20170911-popsignuptwoweeks-perspective_alpha_website_large.jpg');
  transition: transform 1s;
  left: 0;
  top: 0;
}

header:hover:before {
  transform: scale(1.1);
}

h1 {
  color: #F55263;
  margin: 15px 0;
  font-size: 18px;
}

.row {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.row p {
  text-transform: uppercase;
  font-size: 14px;
}

form label {
  text-transform: uppercase;
  margin-top: 50px;
  font-size: 18px;
}

form input {
  display: block;
  width: 97%;
  border-bottom: 1px solid #F55263;
  background-color: rgba(0, 0, 0, 0.7);
  padding: 15px;
  color: white;
  font-size: 15px;
  margin: 20px 0px;
}

.button {
  cursor: pointer;
  margin-left: 20px;
  min-width: 60px;
  padding: 10px;
  text-align: center;
  background-color: #3C3C3C;
  border-radius: 2px;
  text-transform: uppercase;
}

.button-small {
  font-size: 14px;
  padding: 5px;
}

.button.search-button {
  color: white;
  font-size: 18px;
  background-color: #F55263;
  margin-left: auto;
  width: 180px;
}

.selected {
  background-color: #F55263;
}

.container {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
}

.no-content {
  display: flex;
  height: 500px;
  align-items: center;
  justify-content: center;
}

.no-content p {
  color: #C8C8C8;
  font-size: 70px;
}

.toolbar {
  background-color: #F5F5F5;
  padding: 20px 0;
}