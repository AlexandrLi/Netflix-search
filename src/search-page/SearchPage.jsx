import React from 'react';
import { Link } from 'react-router-dom'
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { fetchMovies, reOrder, sortMovies } from '../actions';

import { FilmCard } from '../film-card/FilmCard'
import { Toolbar } from '../toolbar/toolbar'
import s from './search-page.css';

export class SearchPage extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      query: this.props.match.params.query || '',
      order: 'desc'
    }
    this.handleChange = this.handleChange.bind(this);
  }

  toggleOrder(order) {
    this.setState({ order });
    if (this.props.movies.length) {
      this.props.reOrder(order);
    }
  }

  handleChange(event) {
    this.setState({ query: event.target.value });
  }

  toggleSortingType(type) {
    this.props.sortMovies(type, this.state.order);
  }

  static fetchData(dispatch, match) {
    if (match.params.query) {
      dispatch(fetchMovies(match.params.query, 'desc'));
    }
  }

  componentDidMount() {
    if (this.state.query) {
      this.props.fetchMovies(this.state.query, this.state.order);
    }
  }

  onSearch(event) {
    event.preventDefault();
    if (!this.state.query.trim()) {
      return;
    } else {
      this.props.history.push(`/search/${this.state.query}`);
      this.props.fetchMovies(this.state.query, this.state.order);
    }
  }

  render() {
    return (
      <div>
        <header>
          <h1>netflixroulette</h1>
          <form onSubmit={this.onSearch.bind(this)}>
            <label htmlFor="search">find your movie</label>
            <input id="search" type="text" value={this.state.query} onChange={this.handleChange} />
            <div className={s.row}>
              <p>order:</p>
              <div className={`${s.button} ${s.buttonSmall} ${this.state.order === 'asc' ? s.selected : ''}`} onClick={() => this.toggleOrder('asc')}>asc</div>
              <div className={`${s.button} ${s.buttonSmall} ${this.state.order === 'desc' ? s.selected : ''}`} onClick={() => this.toggleOrder('desc')}>desc</div>
              <button className={`${s.button} ${s.searchButton}`}>search</button>
            </div>
          </form>
        </header>
        <div className={s.toolbar}>
          {!!this.props.movies.length && <Toolbar count={this.props.movies.length} sortBy={this.props.sortBy} onSortClick={this.toggleSortingType.bind(this)} />}
        </div>
        {this.props.movies.length ?
          <div className={s.container}>
            {this.props.movies.map(item => <Link to={`/film/${item.id}`} key={item.id}><FilmCard key={item.id} movie={item} /></Link>)}
          </div> :
          <div className={s.noContent}>
            <p>No films found</p>
          </div>}
      </div >
    )
  }
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ fetchMovies, reOrder, sortMovies }, dispatch);
}

function mapStateToProps({ movies, searchType, sortBy }) {
  return { movies, searchType, sortBy };
}

export default connect(mapStateToProps, mapDispatchToProps)(SearchPage);