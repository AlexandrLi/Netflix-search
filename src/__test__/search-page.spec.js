import React from "react"
import renderer from "react-test-renderer"
import { MemoryRouter } from 'react-router-dom'
import { shallow } from "enzyme";
import { SearchPage } from "../search-page/SearchPage";
import { movies } from './mock-data';

const props = {
  movies: movies,
  match: {
    params: {
      query: 'lord of the rings'
    }
  },
  fetchMovies: jest.fn()
}

describe("Search-page snapshot", () => {
  it("Take search-page snapshot", () => {
    const renderedSearchPage = renderer.create(<MemoryRouter><SearchPage {...props} /></MemoryRouter>).toJSON();
    expect(renderedSearchPage).toMatchSnapshot();
  })
});

describe("Shalow render of Search-page component", () => {
  let component;
  beforeEach(() => {
    component = shallow(<MemoryRouter><SearchPage {...props} /></MemoryRouter>)
  });

  it("Component successfully rendered", () => {
    expect(component.length).toEqual(1);
  });
});