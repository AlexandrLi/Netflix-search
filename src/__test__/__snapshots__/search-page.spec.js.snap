// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search-page snapshot Take search-page snapshot 1`] = `
<div>
  <header>
    <h1>
      netflixroulette
    </h1>
    <form
      onSubmit={[Function]}
    >
      <label
        htmlFor="search"
      >
        find your movie
      </label>
      <input
        id="search"
        onChange={[Function]}
        type="text"
        value="lord of the rings"
      />
      <div
        className="row"
      >
        <p>
          order:
        </p>
        <div
          className="button buttonSmall "
          onClick={[Function]}
        >
          asc
        </div>
        <div
          className="button buttonSmall selected"
          onClick={[Function]}
        >
          desc
        </div>
        <button
          className="button searchButton"
        >
          search
        </button>
      </div>
    </form>
  </header>
  <div
    className="toolbar"
  >
    <div
      className="container"
    >
      <div>
        3
         movies found
      </div>
      <div
        className="row"
      >
        Sort By
        <div
          className=""
          onClick={[Function]}
        >
          release date
        </div>
        <div
          className=""
          onClick={[Function]}
        >
          rating
        </div>
      </div>
    </div>
  </div>
  <div
    className="container"
  >
    <a
      href="/film/122"
      onClick={[Function]}
    >
      <div
        className="card"
      >
        <img
          alt="no image"
          src="https://image.tmdb.org/t/p/w342/uexxR7Kw1qYbZk0RYaF9Rx5ykbj.jpg"
        />
        <div
          className="row"
        >
          <h2>
            The Lord of the Rings: The Return of the King
          </h2>
          <div>
            <div
              className="year"
            >
              2003
            </div>
          </div>
        </div>
        <p>
          8.2
        </p>
      </div>
    </a>
    <a
      href="/film/120"
      onClick={[Function]}
    >
      <div
        className="card"
      >
        <img
          alt="no image"
          src="https://image.tmdb.org/t/p/w342/bxVxZb5O9OxCO0oRUNdCnpy9NST.jpg"
        />
        <div
          className="row"
        >
          <h2>
            The Lord of the Rings: The Fellowship of the Ring
          </h2>
          <div>
            <div
              className="year"
            >
              2001
            </div>
          </div>
        </div>
        <p>
          8.1
        </p>
      </div>
    </a>
    <a
      href="/film/121"
      onClick={[Function]}
    >
      <div
        className="card"
      >
        <img
          alt="no image"
          src="https://image.tmdb.org/t/p/w342/wf3v0Pn09jnT5HSaYal7Ami3bdA.jpg"
        />
        <div
          className="row"
        >
          <h2>
            The Lord of the Rings: The Two Towers
          </h2>
          <div>
            <div
              className="year"
            >
              2002
            </div>
          </div>
        </div>
        <p>
          8
        </p>
      </div>
    </a>
  </div>
</div>
`;
